Add the <search> element

domenic opened the pull request:
Closes #5811.

### Mini-explainer

This new element allows authors to express the semantics of "a set of form controls or other content related to performing a search or filtering operation".

This is an especially desired semantic because it is something that an ARIA landmark role exists for (`role="search"`), but today can only be expressed with ARIA. Adding a dedicated element allows authors to follow the [first rule of ARIA](https://w3c.github.io/using-aria/#rule1) (~ "don't use ARIA if you can avoid it") and is part of the [co-evolution process](https://w3c.github.io/aria/#co-evolution) between ARIA and HTML.

Please see [the pull request preview](https://whatpr.org/html/7320/grouping-content.html#the-search-element) for web-developer focused details on usage and examples.

Some notable design choices:

- The name. We were initially a bit concerned that developers would think that this was a type of search input, not a search container. Although we briefly discussed alternative names to combat this like `<searcharea>`, `<searchform>`, `<searchgroup>`, `<searchset>`, they each had problems, and a quick [Twitter poll](https://twitter.com/domenic/status/1437891692926017537) helped seal the deal that developers would expect and prefer the name `<search>`. (In addition to other developers chiming in with reasoning on the thread.) See the comments starting at https://github.com/whatwg/html/issues/5811#issuecomment-916489175 for more details.

- Like other "block-level" elements, `<search>` closes any open `<p>` tags. This ensures an easy transition path from `<div role="search">` to `<search>` and maintains platform consistency.

- Like other "block-level" elements, the user-agent stylesheet for `<search>` makes it `display: block` and `unicode-bidi: isolate`.

- For conformance-checker purposes, the element is categorized as flow content and palpable content, but not sectioning content. (In this respect, it is categorized similarly to `<main>`, but not like `<nav>` or `<aside>`.) We are not sure whether this is ideal, partly because due to #83 sectioning content is a tricky category. But we believe that the main effect of sectioning content is to encourage authors to add headings inside the sections, which is perhaps not the right default here?

- This element is a straightforward replacement for `<div role="search">`, and does not come with any additional functionality (such as having form-submission capabilities). Instead it can be combined with other primitives such as `<form>`, or JavaScript, etc. You can see extensive discussion of this in #5811. The reasons for disfavoring more complex solutions are [implementer-expressed considerations about complexity and compat risk](https://github.com/whatwg/html/issues/5811#issuecomment-1012650363), as well as [web-developer and accessibility community preferences](https://github.com/whatwg/html/issues/5811#issuecomment-981747584).

### Credits

The element description and examples is almost entirely written by @scottaohara, although I made a few edits and rearrangements.

### PR merge checklist

- [x] At least two implementers are interested (and none opposed):
   * Chromium is interested
   * Gecko: positive: https://github.com/mozilla/standards-positions/issues/610
- [x] [Tests](https://github.com/web-platform-tests/wpt) are written and can be reviewed and commented upon at:
   * Parser: https://github.com/web-platform-tests/wpt/pull/39158
   * Rendering: https://github.com/web-platform-tests/wpt/pull/39163
   * Perhaps the conformance checker test suite should also be updated, /cc @sideshowbarker.
- [x] [Implementation bugs](https://github.com/whatwg/meta/blob/main/MAINTAINERS.md#handling-pull-requests) are filed:
   * Chrome: https://bugs.chromium.org/p/chromium/issues/detail?id=1294294
   * Firefox: https://bugzilla.mozilla.org/show_bug.cgi?id=1824121
   * Safari: https://bugs.webkit.org/show_bug.cgi?id=254327
   * Validator.nu: https://github.com/validator/validator/issues/1569
- [x] Dependent specs are ready to update:
  - [x] https://github.com/w3c/html-aria/pull/401
  - [x] https://github.com/w3c/html-aam/pull/355

(See [WHATWG Working Mode: Changes](https://whatwg.org/working-mode#changes) for more details.)


<!--
    This comment and the below content is programmatically generated.
    You may add a comma-separated list of anchors you'd like a
    direct link to below (e.g. #idl-serializers, #idl-sequence):

    Don't remove this comment or modify anything below this line.
    If you don't want a preview generated for this pull request,
    just replace the whole of this comment's content by "no preview"
    and remove what's below.
-->
***
<a href="https://whatpr.org/html/7320/dom.html" title="Last updated on Mar 23, 2023, 1:11 PM UTC (09486ba)">/dom.html</a>  ( <a href="https://whatpr.org/html/7320/5b01d31...09486ba/dom.html" title="Last updated on Mar 23, 2023, 1:11 PM UTC (09486ba)">diff</a> )
<a href="https://whatpr.org/html/7320/grouping-content.html" title="Last updated on Mar 23, 2023, 1:11 PM UTC (09486ba)">/grouping-content.html</a>  ( <a href="https://whatpr.org/html/7320/5b01d31...09486ba/grouping-content.html" title="Last updated on Mar 23, 2023, 1:11 PM UTC (09486ba)">diff</a> )
<a href="https://whatpr.org/html/7320/index.html" title="Last updated on Mar 23, 2023, 1:11 PM UTC (09486ba)">/index.html</a>  ( <a href="https://whatpr.org/html/7320/5b01d31...09486ba/index.html" title="Last updated on Mar 23, 2023, 1:11 PM UTC (09486ba)">diff</a> )
<a href="https://whatpr.org/html/7320/indices.html" title="Last updated on Mar 23, 2023, 1:11 PM UTC (09486ba)">/indices.html</a>  ( <a href="https://whatpr.org/html/7320/5b01d31...09486ba/indices.html" title="Last updated on Mar 23, 2023, 1:11 PM UTC (09486ba)">diff</a> )
<a href="https://whatpr.org/html/7320/parsing.html" title="Last updated on Mar 23, 2023, 1:11 PM UTC (09486ba)">/parsing.html</a>  ( <a href="https://whatpr.org/html/7320/5b01d31...09486ba/parsing.html" title="Last updated on Mar 23, 2023, 1:11 PM UTC (09486ba)">diff</a> )
<a href="https://whatpr.org/html/7320/rendering.html" title="Last updated on Mar 23, 2023, 1:11 PM UTC (09486ba)">/rendering.html</a>  ( <a href="https://whatpr.org/html/7320/5b01d31...09486ba/rendering.html" title="Last updated on Mar 23, 2023, 1:11 PM UTC (09486ba)">diff</a> )
<a href="https://whatpr.org/html/7320/syntax.html" title="Last updated on Mar 23, 2023, 1:11 PM UTC (09486ba)">/syntax.html</a>  ( <a href="https://whatpr.org/html/7320/5b01d31...09486ba/syntax.html" title="Last updated on Mar 23, 2023, 1:11 PM UTC (09486ba)">diff</a> )

KittyGiraudel commented:
Hiiii! ✨  That’s pretty neat—I had no idea this was in the work.

I have a couple questions/comments if that’s okay?
- Is the use of `search` limited to one per page, or can it be used multiple times? The document doesn’t appear to explicitly clarify this.
- In the case it can in fact be used several times on a page, how would a given occurrence be differentiated from another for assistive technologies? 

I guess where I’m coming at is that if it is _not_ unique per page (unlike `<main>`), maybe it should be a sectioning element so that it can essentially be labelled? That’s generally what’s recommended for `<nav>` elements, where we’d encourage authors to label them properly so multiple navigations can be differentiated.

Unless I’m missing something? :)

scottaohara commented:
thanks for the comment, @KittyGiraudel.  There is definitely guidance that can be added to flesh that out some more.

to quickly answer your questions though:
1. this element is different than `<main>`, as there can be more than one per document - though as with any landmark that exposes a landmark role, they should be used with purpose.  Will draft up some guidance for this.
2. With native HTML alone, the `title` attribute would be used to name the element, as there is no other mechanism available at this time without using ARIA.

domenic commented:
I do think the comparison with `<nav>` is worth digging into though. I guess you can have multiple of each per page, and the question is whether we want to encourage `<hN>` children inside of them.

IIUC, for both `<nav>` and `<search>`, per https://w3c.github.io/html-aam/#section-elements-and-grouping-content-elements-not-listed-elsewhere you can only use `title=""` or ARIA. Using a `<hN>` will not suffice to give an accessible name.

Not sure where that leaves us...

scottaohara commented:
have some additional content to add.  my initial PR got messed up though, so will give it a shot again later.

Kaleidea commented on source:
```diff
@@ -19936,7 +19938,82 @@ included with Exhibit B.
+   <dd>Uses <code>HTMLElement</code>.</dd>
+  </dl>
+
+  <p>The <code>search</code> element <span>represents</span> a part of a document or application
```
`<span>represents</span>` might be a leftover `<span>`?

Kaleidea reviewed (changes requested):
<details>
<summary>(edited by @domenic) Off-topic: see discussion in <a href="https://github.com/whatwg/html/issues/5811#issuecomment-965201571">#5811 (comment)</a></summary>
I'd like to thank the editors' work on this proposal and their long-term service towards the web developer community. 
However, [there is a viable alternative](https://github.com/whatwg/html/issues/7323) with improved API ergonomy that better serves developer needs, but wasn't considered.
In my opinion ignoring that alternative would be a missed opportunity to improve developer satisfaction and the quality of the World Wide Web. 

### Opposition

Since this is a low priority feature, there is no need to rush to a solution without giving due consideration for all needs and use-cases, therefore **I have to oppose this proposal in this form at this time**. Due to discourtesy and the [rejection](https://github.com/whatwg/html/issues/5811#issuecomment-964879335) to discuss the alternative of aliasing `<form>`, I've **appealed the decisions to the Steering Group**. I hope their guidance will help achieve an outcome that serves more developers' needs.
</details>

Kaleidea commented on source:
```diff
@@ -19936,7 +19938,82 @@ included with Exhibit B.
+      &lt;li>
+        &lt;p>&lt;a href="services/consulting">Consulting services&lt;/a>&lt;/p>
+        &lt;p>
+          Find out how can help you improve your business with our integrated consultants, Bob and Bob.
```
Word is missing? "how **we** can help"

zcorpan commented:
cc @whatwg/html-parser for

> Does not close `<p>` elements, i.e. does not require any parser changes. (Note: this is unlike `<main>`.) Implementers are generally not excited about parser changes and I don't think auto-closing `<p>` elements is a big enough win.

Is it OK for new "block" elements to not close `<p>`?

If it is, maybe we need to reconsider if `</p>` should be optional, since it will be very confusing to not be able to omit it inside or before `<search>`.

artalar commented:
It would be nice to have `main` property for this element, which indicate to browser that it should be used as a main search for site when user click <kbd>ctrl</kbd>(<kbd>command</kbd>)+<kbd>F</kbd>. It may dramatically improve user search experience for sites with virtual lists.

Kaleidea commented:
> It would be nice to have main property for this element, which indicate to browser that it should be used as a main search for site when user click ctrl(command)+F. It may dramatically improve user search experience for sites with virtual lists.

@artalar  I think that would be a separate proposal to hook and redirect the browser's CtrlCmd-F search functionality to the marked search element on the page, it's more than what this proposal aims for. Currently CtrlCmd-F search can be hooked with JS.

Kaleidea commented:
> cc @whatwg/html-parser for
> 
> > Does not close `<p>` elements, i.e. does not require any parser changes. (Note: this is unlike `<main>`.) Implementers are generally not excited about parser changes and I don't think auto-closing `<p>` elements is a big enough win.
> 
> Is it OK for new "block" elements to not close `<p>`?

@zcorpan  Adding the <search> element requires 2 changes to the WebKit & Blink parsers in the HTMLTreeBuilder class:
[commit](https://github.com/Kaleidea/chromium/commit/f201b98e2de7a522f14d9abdbbb1dd5a5a85b0c6#diff-cd7d1757e6ad8e83535f50a5c090c98dc62a477ff943b15a645d1e3c31caf268).

Making it a paragraph closing element requires 2 more trivial changes:
[commit](https://github.com/Kaleidea/chromium/commit/0a15a4c15bd952bc3b439d2607839a3942d42665).


OTOH if `<search>` is not closing paragraphs then the trivial transformation from `<div role=search>` -> `<search>` will break in cases like:
`<p> Paragraph ends here. <div role=search> ... </div>`
will not be equivalent to:
`<p> Paragraph ends here. <search> ... </search>`
Such inconsistency can be a difficulty for dev tool developers (eg. automated migration to `<search>`) and a footgun for web developers.

Note: I don't like the optional `</p>` either, but changing it would break the web.

zcorpan commented:
> OTOH if `<search>` is not closing paragraphs then the trivial transformation from `<div role=search>` -> `<search>` will break in cases like: `<p> Paragraph ends here. <div role=search> ... </div>` will not be equivalent to: `<p> Paragraph ends here. <search> ... </search>` Such inconsistency can be a difficulty for dev tool developers (eg. automated migration to `<search>`) and a footgun for web developers.

I agree, and it will be a persistent footgun. I think it's probably better in the long run to change the parser when adding a new block element.

domenic commented:
@hsivonen any updates on your thoughts about whether it's OK to have `<search>` close `<p>` tags?

hsivonen commented:
If a `<search>` element is added, making it close `<p>` is probably the right thing to do. I'm not convinced that adding the `<search>` element is worthwhile, since it doesn't add expressiveness but mainly addresses a principled concern that all of ARIA should be ported to HTML in non-ARIA form, but I don't know enough about the use cases to argue strongly in favor or agains..

domenic commented:
Alright, I've added the `<p>` tag closing, and updated the OP with a summary of the state of things. I've additionally filed for TAG review at https://github.com/w3ctag/design-reviews/issues/714 .

My guess on the next roadblock will be writing web platform tests, since per https://github.com/web-platform-tests/wpt/issues/27868 any tests involving the parser are annoying. But, it'd still be ideal to get editor review of this PR, and maybe prepare the PRs discussed in https://github.com/w3c/html-aria/issues/371 and https://github.com/w3c/html-aam/issues/369, so that we can merge this as soon as those are ready.

Kaleidea commented:
**As the [implementer of this feature *without form functionality*]( https://github.com/Kaleidea/chromium/commits/search-element-div ) in Chromium, I am OPPOSING this specification** until proper user research and origin trial is conducted.

I have put an order of magnitude more [effort]( https://kaleidea.github.io/whatwg-search-proposal/ ) (about 2 months) than all other participants combined into progressing this feature towards an outcome that best meets the web developer community's needs. The current proposal does not meet that purpose.

An [appeal]( https://github.com/whatwg/html/issues/7594 ) was filed according to the [workstream policy]( https://whatwg.org/workstream-policy#appeals ). Due to violations of WHATWG procedures and policies, this specification cannot be merged until those issues are resolved.

@domenic
1. Please register this opposition appropriately in the description.
2. Please add the "do not merge" label.

This is expected to happen within hours. Please respect my opposition and do not attempt to ignore or censor it, as [it happened before]( https://github.com/whatwg/html/pull/7320#pullrequestreview-802253725 ). I appreciate your fairness.

zcorpan commented:
A new element with the same capabilities as `<form>` is a non-starter. The security fallout from that will take decades at best to iron out.

`<form>` is an XSS sink, plus it is able to perform fetches that are more powerful compared to other elements (it can perform `POST` requests etc) and navigation.

A new element that can be a `form` would complicate all of the features that interact with forms, and each of those have a potential to break web content.

The HTML parser has special knowledge about forms (e.g. the form element pointer). Extending that to also apply to a new element seems bad (and I expect is not going to be accepted by API owners for browsers' HTML parsers, as this, in itself, is an XSS-risky change), but also not doing that is somewhat bad because then previously-working page with `form` might break when switching to `search`.

Furthermore, making a new element that *sometimes* act as a `form` (based on the presence of the `action` attribute) is even worse. That would likely be a source of bugs for a long time, if the experience with elements like `object` is any indication.

As for the stated motivation:

> 1. a [popular developer request](https://kaleidea.github.io/whatwg-search-proposal/#requests-for-this-feature)

`<form><search>` or `<form type=search>` seems like it would satisfy the requests without complicating everything and introducing new XSS risk and web compat risk.

> 2. the best practice as demonstrated by the [MDN example](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/Search_role) and the [WAI-ARIA example](https://www.w3.org/TR/wai-aria-practices/examples/landmarks/search.html).

Examples in MDN and APG isn't a compelling argument for how to shape a new feature in my opinion. They show how to use existing features.

> 3. the common sense [mental model](https://search-element-html-spec.netlify.app/multipage/forms.html#the-search-element): “The search element represents a part of a document or application that contains a set of form controls or other content related to performing a search or filtering operation.” A set of form controls is a form.

`<p>` and `<fieldset>` can also represent a set of form controls. It doesn't follow that they should also be forms.

From what I can tell, making a new element act like a `form` violates these HTML design principles:

- https://www.w3.org/TR/html-design-principles/#support-existing-content (changes to legacy APIs, form submission model, the HTML parser, etc, can break existing content)
- https://www.w3.org/TR/html-design-principles/#degrade-gracefully (in legacy UAs users won't be able to submit forms)
- https://www.w3.org/TR/html-design-principles/#do-not-reinvent-the-wheel (`form` already exists)
- https://www.w3.org/TR/html-design-principles/#solve-real-problems (changes to fit a stated mental model, but the same use cases are possible with the alternative smaller and less risky change)
- https://www.w3.org/TR/html-design-principles/#secure-by-design (introduce new XSS risk without introducing new functionality)
- https://www.w3.org/TR/html-design-principles/#avoid-needless-complexity (as explained above)

zcorpan commented:
@domenic

> Alright, I've added the `<p>` tag closing,

You also need to make `<search>` start tag close `p`, here:

> A start tag whose tag name is one of: "address", "article", "aside", "blockquote", "center", "details", "dialog", "dir", "div", "dl", "fieldset", "figcaption", "figure", "footer", "header", "hgroup", "main", "menu", "nav", "ol", "p", "section", "summary", "ul"

Everything else in this PR: LGTM.

domenic commented:
> You also need to make `<search>` start tag close `p`, here:

Wow, today I learned about "button scope", amazing stuff.

> I started this with new tests in html5lib-tests, based on existing tests for the main element: https://github.com/html5lib/html5lib-tests/pull/142

Thank you so much!!

annevk commented on source:
```diff
@@ -20099,7 +20101,114 @@ included with Exhibit B.
 
   </div>
 
+  <h4>The <dfn element><code>search</code></dfn> element</h4>
```
We're eating some newlines around here. I guess that's okay?

annevk reviewed (approved):
Looks good!

scottaohara commented:
was speaking with @mfreed7 on a completely unrelated topic, but it reminded me that landing `<search>` seems to have fallen off the radar.  So, while I'm not sure I can attend, Mason mentioned I should add the agenda+ label so that the next triage meeting a status update on what else may be needed to get this resolved.

thanks!

domenic commented:
This is blocked purely on tests. https://github.com/html5lib/html5lib-tests/pull/142 was merged but getting that into web-platform-tests is a chore: https://github.com/web-platform-tests/wpt/issues/27868 implies we need to update `html5lib-python` and then `wpt`, which nobody been able to do yet. (In general, html5lib-python seems unmaintained, so, that's not great.)

This is very sad :(

scottaohara commented:
Thank you, @domenic.  

So from what you mention, this would be a blocker for any other proposed HTML element as well - i.e., potentially `selectmenu`?

Also, shall I leave the agenda+ label on this, or remove it as it's really more an issue with being blocked by the tests?

domenic commented:
It'd be an issue for any HTML element that changes parser behavior. The most common parser behavior change is auto-closing `<p>` tags. However, I think that doesn't apply to `<selectmenu>`.

agenda+ might be reasonable to leave on, in order to try to guilt someone on the call into doing the work?

annevk commented:
It appears that https://github.com/web-platform-tests/wpt/pull/39158 would unblock this testing-wise.

It seems that https://github.com/w3c/html-aam/issues/369 and corresponding PR cover the accessibility side.

What remains:

* Implementation bugs.

zcorpan commented:
I filed:
   * Firefox: https://bugzilla.mozilla.org/show_bug.cgi?id=1824121
   * Safari: https://bugs.webkit.org/show_bug.cgi?id=254327

zcorpan commented:
OK I think this is ready now, modulo review of the rendering tests.

annevk commented:
I suggest we let @domenic do the honors since it's been well over a year since we started all this and he might well have some other relevant insights.

domenic commented:
Thank you all for helping push this over the finish line!! I think it's ready to merge. I'll also file an MDN bug.
